server/src/http/broadcaster.rs

```diff
@@ -0,0 +1,190 @@
+/// copied from https://github.com/actix/examples/blob/master/server-sent-events/src/broadcast.rs
+use std::{
+    collections::HashMap,
+    hash::{Hash, Hasher},
+    sync::Arc,
+    time::Duration,
+};
+
+use actix_web::{
+    rt::time::interval,
+    web::{Json, Query},
+};
+use actix_web_lab::{
+    sse::{self, Event, Sse},
+    util::InfallibleStream,
+};
+use dashmap::DashMap;
+use futures_util::future;
+use parking_lot::Mutex;
+use serde::Serialize;
+use tokio::sync::mpsc;
+use tokio_stream::wrappers::ReceiverStream;
+use unleash_types::client_features::{ClientFeatures, Query as FlagQuery};
+
+use crate::{
+    filters::{filter_client_features, name_prefix_filter, project_filter, FeatureFilterSet},
+    tokens::cache_key,
+    types::{EdgeToken, FeatureFilters},
+};
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
+struct QueryWrapper {
+    query: FlagQuery,
+}
+
+impl Hash for QueryWrapper {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        serde_json::to_string(&self.query).unwrap().hash(state);
+    }
+}
+
+#[derive(Clone, Debug)]
+struct ClientGroup {
+    clients: Vec<mpsc::Sender<sse::Event>>,
+    filter_set: Query<FeatureFilters>,
+    token: EdgeToken,
+}
+
+#[derive(Default)]
+struct BroadcasterInner {
+    active_connections: HashMap<QueryWrapper, ClientGroup>,
+}
+
+pub struct Broadcaster {
+    inner: Mutex<BroadcasterInner>,
+    features_cache: Arc<DashMap<String, ClientFeatures>>,
+}
+
+impl Broadcaster {
+    /// Constructs new broadcaster and spawns ping loop.
+    pub fn new(features: Arc<DashMap<String, ClientFeatures>>) -> Arc<Self> {
+        let this = Arc::new(Broadcaster {
+            inner: Mutex::new(BroadcasterInner::default()),
+            features_cache: features,
+        });
+
+        Broadcaster::spawn_ping(Arc::clone(&this));
+
+        this
+    }
+
+    /// Pings clients every 30 seconds to see if they are alive and remove them from the broadcast
+    /// list if not.
+    fn spawn_ping(this: Arc<Self>) {
+        tokio::spawn(async move {
+            let mut interval = interval(Duration::from_secs(30));
+
+            loop {
+                interval.tick().await;
+                this.remove_stale_clients().await;
+            }
+        });
+    }
+
+    /// Removes all non-responsive clients from broadcast list.
+    async fn remove_stale_clients(&self) {
+        let active_connections = self.inner.lock().active_connections.clone();
```

Comment:
we obtain the lock only on this line. By the time we get to line 102 someone in a race condition may change the active_connections. Maybe the lock should be in a variable scoped to the whole method?

Reply:
quite possibly! there's also the possibility that there's a much better data structure to use for this 💁🏼

Reply:
When it comes to data_structure. I believe the answer might be in the data structure used in the rest of Edge. The DashMap. It's meant for concurrent work.

Reply:
Yeah, yeah; that's what I was thinking of. But it wasn't used in the example and I didn't wanna start mucking about with it under the time pressure we were looking at last week. I'll see if I can't find some time to try it out before the review sesh tomorrow.